core/src/main/scala/org/apache/spark/scheduler/cluster/ExecutorData.scala

```diff
@@ -38,5 +39,6 @@ private[cluster] class ExecutorData(
     override val totalCores: Int,
     override val logUrlMap: Map[String, String],
     override val attributes: Map[String, String],
-    override val resourcesInfo: Map[String, ExecutorResourceInfo]
+    override val resourcesInfo: Map[String, ExecutorResourceInfo],
+    @volatile var isReady: Boolean = false
```

Comment:
Instead of tracking more state, why not keep `freeCores` at 0 until the message you're adding arrives?